### DISCUSSION TITLE
How does one enable Netbird's DNS feature?

### DISCUSSION THREAD

**Discussion by gene1wood** [netbirdio/netbird#638] [Q&A / Support]
I've read through the documentation ( https://netbird.io/docs/how-to-guides/nameservers ) but it feels like it's missing a section on how to get a client to use this DNS feature.

I get that Netbird spins up a local resolver on the Netbird interface listening on port 53.

> It listens on the peer's IP, and usually, it will use the default port 53, but if it is in use, it will use the 5053 port.

But what is the intended way to get a client to use this local resolver?

The first thing I tried, was going into my network settings on my NIC (in Ubuntu and Windows) and setting the Netbird IP as the DNS resolver.

This works great (I'm able to resolve `netbird.cloud` DNS names) as long as Netbird is already running, but when you reboot, the OS comes up, Netbird isn't running yet and so no DNS resolution works, including the DNS resolution that Netbird needs to be able to start.

So the solution can't be to configure the local resolver in the network stack.

Is the OS meant to be consuming from the Netbird virtual NIC's DHCP response a name server? If so, how is the OS meant to know what to do when it has two network interfaces (the Netbird virtual NIC and the actual ethernet NIC), each of which gets a DHCP response with a DNS server? 

What would the section of the documentation above say about how to actually use the DNS feature, if it included that? I'm specifically curious about how this is meant to be used under Windows 10 and Ubuntu 18/20/22 LTS

**Comment by mlsmaycon** [comment]
Hello @gene1wood, I understand your confusion, there are some notes in the documentation, but we need to improve it. 

The feature works automatically; no extra manual configuration should be required in the peers. Still, there are some things that are not straightforward to visualize. For instance, on Windows, macOS, and Linux with systemd-resolved, you won't see a change on /etc/resolv.conf or in the interface settings because these systems either abstract the DNS resolution with their own resolvers, which allows domain routing (if domain == X use server Y). You can use the following commands to validate the DNS configuration:

Windows (Powershell)
```shell
Get-DnsClientNrptPolicy
```

You should see something similar to:
```shell

Namespace                        : .netbird.cloud
QueryPolicy                      :
SecureNameQueryFallback          :
DirectAccessIPsecCARestriction   :
DirectAccessProxyName            :
DirectAccessDnsServers           :
DirectAccessEnabled              :
DirectAccessProxyType            : NoProxy
DirectAccessQueryIPsecEncryption :
DirectAccessQueryIPsecRequired   : False
NameServers                      : 100.64.0.207
DnsSecIPsecCARestriction         :
DnsSecQueryIPsecEncryption       :
DnsSecQueryIPsecRequired         : False
DnsSecValidationRequired         : False
NameEncoding                     : Utf8WithoutMapping
```

MacOS 
```shell
scutil --dns
```

You should see something similar to:
```shell
resolver #2
  domain   : netbird.cloud
  nameserver[0] : 127.0.0.1
  port     : 53
  flags    : Supplemental, Request A records, Request AAAA records
  reach    : 0x00030002 (Reachable,Local Address,Directly Reachable Address)
  order    : 101800
```

Linux (systemd-resolved) 
```shell
resolvectl dns
resolvectl domain
```

You should see something similar to:
```shell
ubuntu@peer:~$ resolvectl dns
Global:
Link 13 (wt0): 100.64.0.231
ubuntu@peer:~$ resolvectl domain
Global:
Link 13 (wt0): netbird.cloud
```

Another point is the requirement for a nameserver configuration to resolve all domains on Linux without systemd-resolved; this is needed because most systems won't support domain routing, and we decided not to mess with the OS /etc/hosts file for peer name resolution. So, if you run into issues, you can validate if you ran into this case by adding a nameserver with the following configuration:

<img width="376" alt="image" src="https://user-images.githubusercontent.com/7747744/209915741-ef632102-9977-44b5-9db0-017a88c3f7cd.png">


Besides updating the documentation, we plan to add:
1. DNS information to the `netbird status` command
2. Dashboard warning if a `all domains` nameserver is needed for any peer
